### DISCUSSION TITLE
Query for euclidean signed distance (ESDF) from the map?

### DISCUSSION THREAD

**Discussion by balakumar-s** [ethz-asl/wavemap#64] [Q&A] [answered]
Does the map store esdf at every voxel or at least have a way to query for signed distance?

**Comment by balakumar-s** [comment]
The answer is No, from the FAQ: https://ethz-asl.github.io/wavemap/pages/faq.html#does-wavemap-support-euclidean-signed-distance-fields

I also found a new paper that integrates with RMPs: https://events.infovaya.com/presentation?id=121343
Curious on how distance was calculated from the occupancy map.

**Reply by victorreijgwart** [reply, marked as answer]
Hi @balakumar-s,
The main map only stores occupancy. However, there's already full support for batched ESDF computation and experimental support for multi-resolution ESDFs. We will likely add support for incremental ESDF updates in the future.
In general, there are two ways to get the signed distance from a query point to the closest obstacle. For each query point you could either search the closest obstacle and then compute the signed distance, or you could query an ESDF after precomputing it for a region of interest. Which option is best (latency, memory, compute,...) depends on the number (and spatial distribution) of the point samples at which the signed distance is needed. Note that searching neighboring obstacles is fast since the map obstacles are naturally arranged hierarchically, in octrees.
For the paper you mentioned, which integrates wavemap with RMPs, we need the distance and relative orientation w.r.t. all the obstacles within a radius of interest (instead of only the single closest one). I'll summarize the paper here in case it's relevant to you or other people reading this post. In brief, we find the occupied cells within a certain radius around the robot using a hierarchical algorithm. We then compute the distance and relative orientation of the robot w.r.t. each occupied cell directly. In a fixed-resolution setting, this would be too expensive as the worst-case number of obstacle cells grows cubically in both the resolution and the considered radius. However, high-resolution obstacles are only needed in the robot's direct vicinity and, as shown in the paper, distant obstacles can be considered at a lower resolution without affecting the solution. So, by gradually decreasing the resolution as the distance to the robot increases, we can combine a high resolution nearby and a large collision avoidance radius while keeping the worst-case number of considered cells reasonable (e.g. <50k). Compared to optimization based approaches using ESDFs, this method adds no latency and computing overheads from map preprocessing. Although we only considered Euclidean robot-obstacle distances in this paper (similar to ESDFs), the method would support arbitrary distance formulations - basically any collider could be used to represent the robot (multiple primitives, a full BVH,...).
